# sc-atac-seq-pipeline
ATAC-seq pipelines for HuBMAP

# Overview

The HuBMAP Consortium sc-atac-seq pipeline is a three-stage pipeline for
analyzing scATAC-seq data sets, composed of [SnapTools](https://github.com/r3fang/SnapTools "SnapTools"),
[SnapATAC](https://github.com/r3fang/SnapATAC "SnapATAC"),
and [chromVAR](https://bioconductor.org/packages/release/bioc/html/chromVAR.html "chromVAR").

SnapTools performs quantification using a specified aligner,
and HuBMAP has standardized on BWA with the GRCh38 reference genome.
SnapTools divides the genome into non-overlapping bins of user-specified size
(we use 1000, 5000, and 10000) and produces a binary cell-by-bin matrix
denoting whether reads in each cell were aligned to each bin.

The SnapATAC
secondary analysis pipeline filters bins based on the ENCODE genomic region
blacklist, performs t-SNE dimensionality reduction, and selects peaks from
all available bins. The chromVAR tool performs motif analysis, assigns motifs
to transcription factors, and computes differential enrichment of transcription
factors across cells in the data set.

# Inputs
## Required
* sequence_directory

A directory for the pipeline to search for fastq or fastq.gz files. The pipeline
only works on paired end reads and expects, for historical reasons, the paired
end read files to be named
< some name >\*_R1\*.fastq and < some name >\*_R3\*.fastq. If a file
containing barcodes < some name >\*_R2\*.fastq is found the barcodes will be read and added
to the read IDs in the paired end fastq/fastq.gz files

* input_reference_genome

A fasta file of the GRCh38 reference genome

* encode_blacklist

The genome BED file with the blacklisted regions to be removed.
E.g hg38.blacklist.bed

* gene_track

A BED file of gene tracks. E.g gencode.v32.annotation.bed

## Optional

* reference_genome_index

A .gz file containing the BWA generated index of the GRCh38 reference genome.
I.e. the ".bwt", ".sa", ".ann", ".pac", ".amb" files generated by BWA
indexing. If this file is provided the index will not have to be generated
by the pipeline saving some time.

* gene_annotation

A GTF file of gene annotations. If there is no preferred_barcodes CSV file
 cells will be selected based on "coverage" and "promoter ratio -"
i.e. the percentage of reads mapped to promoters per cell

* preferred_barcodes

Cells will be selected based on the barcodes listed in the CSV file. E.g
[barcodes CSV](http://renlab.sdsc.edu/r3fang/share/github/Mouse_Brain_10X/atac_v1_adult_brain_fresh_5k_singlecell.csv
 "barcode CSV")

* promoters

If no no barcode CSV or gene annotation file is provided this option is
required. (this requirement will be removed)


# Outputs

* Bins.csv

A CSV file providing sequence name and bin information

* cellBarcodes.CSV

A CSV file with barcode ID and barcode

* cellByBin_summary.csv

A CSV file with barcode ID and bin number

* cellClusterAssignment.csv

A CSV file with barcode ID and cluster number

* GenesRanges.csv

A CSV file providing sequence, gene name and gene location information

* cellByGene.mtx

A file with the cell by gene matrix in Matrix Market format

* cellGenes.csv

A CSV file with gene ID and gene name

* peaksAllCells.csv

A CSV file with sequence name and peak start and end
